Code/GraphMol/RWMol.cpp

```diff
@@ -574,26 +576,274 @@ void RWMol::beginBatchEdit() {
+                continue;
+            }
+            Bond *obnd = getBondBetweenAtoms(aid1, oIdx);
+            
+            // See if the bond is already gone or if we are going to delete the nbrbond anyway.
+            if (!obnd || delBonds[obnd->getIdx()]) {
+                continue;
+            }
+            if (std::find(obnd->getStereoAtoms().begin(), obnd->getStereoAtoms().end(),
+                          aid2) != obnd->getStereoAtoms().end()) {
+                // github #6900 if we remove stereo atoms we need to remove
+                //  the CIS and or TRANS since this requires stereo atoms
+                if (obnd->getStereo() == Bond::BondStereo::STEREOCIS ||
+                    obnd->getStereo() == Bond::BondStereo::STEREOTRANS) {
+                    obnd->setStereo(Bond::BondStereo::STEREONONE);
+                }
+                obnd->getStereoAtoms().clear();
+            }
+        }
+        boost::tie(a1, a2) = boost::adjacent_vertices(aid2, d_graph);
+        while (a1 != a2) {
+            auto oIdx = rdcast<unsigned int>(*a1);
+            ++a1;
+            if (oIdx == aid1) {
+                continue;
+            }
+            Bond *obnd = getBondBetweenAtoms(aid2, oIdx);
+            // See if the bond is already gone or if we are going to delete the nbrbond anyway.
+            if (!obnd || delBonds[obnd->getIdx()]) {
+                continue;
+            }
+            if (std::find(obnd->getStereoAtoms().begin(), obnd->getStereoAtoms().end(),
+                          aid1) != obnd->getStereoAtoms().end()) {
+                // github #6900 if we remove stereo atoms we need to remove
+                //  the CIS and or TRANS since this requires stereo atoms
+                if (obnd->getStereo() == Bond::BondStereo::STEREOCIS ||
+                    obnd->getStereo() == Bond::BondStereo::STEREOTRANS) {
+                    obnd->setStereo(Bond::BondStereo::STEREONONE);
+                }
+                obnd->getStereoAtoms().clear();
+            }
+        }
+        
+        removeSubstanceGroupsReferencingBond(*this, idx);
+        
+        bnd->setOwningMol(nullptr);
+        
+        MolGraph::vertex_descriptor vd1 = boost::vertex(aid1, d_graph);
+        MolGraph::vertex_descriptor vd2 = boost::vertex(aid2, d_graph);
+        boost::remove_edge(vd1, vd2, d_graph);
+        delete bnd;
+        --numBonds;
+    }
+    
+    // loop over all bonds with higher indices than the minimum modified and update their indices
+    ROMol::EDGE_ITER firstB, lastB;
+    boost::tie(firstB, lastB) = this->getEdges();
+    unsigned int next_idx = min_idx;
+    while (firstB != lastB) {
+        Bond *bond = (*this)[*firstB];
+        if (bond->getIdx() > min_idx) {
+            bond->setIdx(next_idx++);
+        }
+        ++firstB;
+    }
+}
+
+void RWMol::batchRemoveAtoms() {
+    if(!dp_delAtoms)
+        return;
+    std::vector<Atom*> oldIndices;
+    oldIndices.reserve(getNumAtoms());
```

Comment:
This could be optimized with an alloca or std::vector style allocator.

Reply:
and it needs to be. At the moment you reserve the memory, but you never actually resize the vector and then you assign using `operator[]`. This is UB

Reply:
That was certainly braindead.